### DISCUSSION TITLE
dose  zustand support latest vesion React 19？

### DISCUSSION THREAD

**Discussion by BertramYe** [pmndrs/zustand#2686] [Q&A] [answered]
hello when I use the latest React19 of the Next 15 RC , I find the error here below , but it's OK  in the React 18:

```bash

npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR!
npm ERR! While resolving: zustand@4.5.5
npm ERR! Found: react@19.0.0-rc-f994737d14-20240522
npm ERR! node_modules/react
npm ERR!   peer react@"19.0.0-rc-f994737d14-20240522" from next@15.0.0-rc.0 
npm ERR!   node_modules/next
npm ERR!     peer next@"^14.0.0-0 || ^15.0.0-0" from next-auth@5.0.0-beta.20
npm ERR!     node_modules/next-auth
npm ERR!       next-auth@"^5.0.0-beta.20" from the root project
npm ERR!     next@"15.0.0-rc.0" from the root project
npm ERR!   peer react@"^18.2.0 || ^19.0.0-0" from next-auth@5.0.0-beta.20   
npm ERR!   node_modules/next-auth
npm ERR!     next-auth@"^5.0.0-beta.20" from the root project
npm ERR!   3 more (react-dom, styled-jsx, the root project)
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peerOptional react@">=16.8" from zustand@4.5.5
npm ERR! node_modules/zustand
npm ERR!   zustand@"4.5.5" from the root project
npm ERR!
npm ERR! Conflicting peer dependency: react@18.3.1
npm ERR! node_modules/react
npm ERR!   peerOptional react@">=16.8" from zustand@4.5.5
npm ERR!   node_modules/zustand
npm ERR!     zustand@"4.5.5" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR!
npm ERR!
npm ERR! For a full report see:
npm ERR! C:\Users\Administrator\AppData\Local\npm-cache\_logs\2024-08-17T10_56_00_869Z-eresolve-report.txt

npm ERR! A complete log of this run can be found in: C:\Users\Administrator\AppData\Local\npm-cache\_logs\2024-08-17T10_56_00_869Z-debug-0.log


```
![image](https://github.com/user-attachments/assets/f590811d-89d1-40d1-9dc4-82ee816022f6)

**Comment by dai-shi** [comment]
We won't get such errors once React 19 is out.
Until then, you need `--force`.

**Reply by BertramYe** [reply]
🆗， but I am afraid that I can't use the zustand in the React 19 , because even I install  the latest package of zustand in React 19 forcely, but I can't use it as there were  errors happend when in the React 19 with the same code of the React 18, where it works fine.

**Reply by dai-shi** [reply]
> as there were errors happend when in the React 19

What's the error? Do you mean the npm error? I'm not sure how controllable, but you may need to patch zustand on your end if `--force` isn't acceptable. (Or, maybe we support prereleases. btw, I didn't know Next exposes rc versions. How about Next 14?)

**Reply by BertramYe** [reply]
no, I mean the Next 15rc, which is required the React 19 rc, and Next 14 is React 18, and also I have trid with the React-Redux, it seems there were same error with the installation , but when I install with the `--legacy-peer-deps` in the  Next 15rc with the React 19 rc, no error duiring the installation, but when I use the store in the projects, error happend like here below, but when I use in the Next 14 with the same code , no error.
![image](https://github.com/user-attachments/assets/30ff7bfa-4ecf-40bc-80b3-1f9934da9e06)

**Reply by dai-shi** [reply]
I see. Can you share the minimal reproduction code that fails in Next 15-rc and works in Next 14? It feels like it's not an issue in Zustand. It would be better if you can reproduce the error without Next 15-rc, but only with React 19-rc. StackBlitz preferred.

**Reply by BertramYe** [reply]
sure, here are the store codes,
![image](https://github.com/user-attachments/assets/2c0272d7-3a38-46ee-9290-643b23929914)
as you can see, when I fetched the datas via the api, it's ok, but when I get the datas with the codes below, error happend,
![image](https://github.com/user-attachments/assets/5169334a-3612-4805-8582-16bd0fe56cfe)
![image](https://github.com/user-attachments/assets/2c64b833-af82-460b-83d3-fe65e673eacc)

**Reply by dai-shi** [reply]
I can't reproduce the error with the screenshots, but anyway unless you can create a _minimal_ (= as few lines as possible) repro, can't help either.
The screenshot says it violates the rule of hooks. You want to solve it first.

**Reply by BertramYe** [reply]
🆗， here are mini codes , as you can see here below, may be you can re-pro the errors:

```tsx
//  store
import { create } from "zustand";

type BlogType = {
    [key: string]: string
}



type response = {
    success:boolean,
    data?: BlogType[],
    error?: string;
}

const GetBlogs =  async () => {
    const blogResult:response =  {
        success:false
    }
    try{
        const blogs = [
            {
                id:"1",
                title:"this is test datas1,this is test datas1,this is test datas1,this is test datas1",
                desc: "this is test datas1",
            },
            {
                id:"2",
                title:"this is test datas2",
                desc: "this is test datas2 content",
            },
            
        ]
        const searchResult = await new Promise(resolve=>{
            resolve(blogs)
        })
        blogResult.data = searchResult as BlogType[]
    }catch(error){
        blogResult.error = "search blogs failed !"
    }

    return blogResult
}



type BlogsStore = {
    blogs: BlogType[],
    updateBlogs:(newBlogsList:BlogType[])=>void,
    fetchAllBlogs: ()=> Promise<void>
}



const blogStore = create<BlogsStore>()((set) => (
    {
        blogs: [],
        updateBlogs: (newBlogsList:BlogType[]) => set((state) => ({blogs: newBlogsList})),
        // get all datas
        fetchAllBlogs: async () => {
            const response = await GetBlogs();
            set({ blogs: response.data? response.data:[] });
        }
    }
))


export default blogStore

```

as you can see , when I calling the promise in the `useEffect ` hooks, the error occurred, so the zustand's promise can't be calling in the `useEffect` here below?

```tsx
// use in the test page 

'use client'
import { blogStore } from "@/store"
import { useEffect, useRef } from 'react';



const TestPage = () => {
    const { blogs, fetchAllBlogs } = blogStore()
    const showblogRef = useRef(null)
    
    useEffect(()=>{
        if(showblogRef.current){
            fetchAllBlogs()      // this steps will result the error
        }
    },[])


    return (
        <div>
            <div ref={showblogRef}>{blogs.length}</div>
        </div>
    )
}

export default TestPage


```

**Reply by dai-shi** [reply]
copy pasted to StackBlitz. Is it working as expected for now?
https://stackblitz.com/edit/vitejs-vite-e57amf?file=src%2FApp.tsx&terminal=dev

**Reply by BertramYe** [reply]
ok, look, your React's version is 18.* it works fine, when I use the React 18 , the error happend when using the React 19 rc
![image](https://github.com/user-attachments/assets/4cb7620a-5b90-4c11-ae16-507d3ebd5fd2)

here are my React version here , can you have a tried with the re-pro?:
![image](https://github.com/user-attachments/assets/97c84ef3-b87a-4fad-bc77-abe57bbb16a7)

**Reply by dbritto-dev** [reply]
@BertramYe I believe you have two react versions, zustand use React v18 but since you are installing React v19rc probably you have two versions of React. Would you mind using `overrides` to specify React v19rc as the right version?

**Reply by dai-shi** [reply]
Yes, it was React 18. I wanted to make sure if it's on the right track.

Feel free to fork the StackBlitz project.
Here's the one with `react@rc`. https://stackblitz.com/edit/vitejs-vite-kukux6?file=package.json%3AL14&terminal=dev How is it?

**Reply by BertramYe** [reply]
em....per test, I should found the root reason, it was caused by the reactCompiler , 
 and also here are the  Nextjs15  offical Doc(https://nextjs.org/blog/next-15-rc), and when I open and use the react compiler in the nextjs 15 rc with the zustand  , the error happend,
![image](https://github.com/user-attachments/assets/b55215c3-79a5-4704-ac9f-b2054992ce4b)

and also when I close the reactCompiler in the NextJS 15 RC, it works fine, also here are I have create a project demo(https://stackblitz.com/edit/stackblitz-starters-uajzp6?file=next.config.js), you can also re-pro the error with the config ( reactCompiler: true,) and as for the pure react 19 rc, it should works fine without the React compiler(as the forked projects you shared me and I have also confirmed ), but I don't know weather it has the same error with the React compiler , I need have do more test, I think.

**Reply by dai-shi** [reply]
https://stackblitz.com/edit/vitejs-vite-xzr6i9?file=vite.config.ts,src%2FApp.tsx&terminal=dev enables React Compiler, but is it working?

**Reply by BertramYe** [reply]
according to the offical docs (https://react.dev/learn/react-compiler), to start the react compiler in the vite, shouldn't it add the config here below?

```
// vite.config.js
const ReactCompilerConfig = { /* ... */ };

export default defineConfig(() => {
  return {
    plugins: [
      react({
        babel: {
          plugins: [
            ["babel-plugin-react-compiler", ReactCompilerConfig],
          ],
        },
      }),
    ],
    // ...
  };
});
```
and also just add it , it seems it can't work normally in the codes : https://stackblitz.com/edit/vitejs-vite-gmxidb?file=vite.config.ts

![image](https://github.com/user-attachments/assets/a2ec8b6f-84e0-415f-acb4-9fda33d67505)
 
no result and no error show here below
![image](https://github.com/user-attachments/assets/c74b3eb1-3e9e-4cf0-ade4-12f43f4d39d5)

**Reply by dai-shi** [reply]
> shouldn't it add the config here below?

oh, you are right. that's my mistake.

**Reply by BertramYe** [reply]
also in the nextjs, it did cause the error like this with the same codes(https://stackblitz.com/edit/stackblitz-starters-idveqs?file=next.config.js):
![image](https://github.com/user-attachments/assets/80aa2d95-7963-4072-a371-7e2b386b3a27)

**Reply by dai-shi** [reply]
And, here's the fix: https://stackblitz.com/edit/vitejs-vite-vbtlq7?file=vite.config.ts,src%2FApp.tsx%3AL49

**Reply by BertramYe** [reply]
yeah, it works now, it seem just because the hooks name ? it must start with the ^use ? I found when I change the  hooks name from 'blogStore ' into 'useBlogStore', the error fixed in the NextJS 15rc as well

**Reply by dai-shi** [reply, marked as answer]
Yeah, React Compiler would only recognize hooks with the name. I think you can reproduce the same error without Zustand.

**Reply by BertramYe** [reply]
ok, Great, I will try later , but anyway, thank you very much ❤,and maybe if possible, you can note this point in your offical doc to avoid the same error occurred in the future when use the zustand, I think 😀.

**Reply by dbritto-dev** [reply]
@BertramYe that's not related to Zustand but React itself, so you should try not to break React rules

**Reply by dai-shi** [reply]
Agree with @dbritto-dev.
Besides, we never teach to remove `use` prefix from hook name. (Let me know why you named like `blogStore`. Then, we might be able to improve our docs. "use" prefix is important for eslint rules to work.)
I think React Compiler (and React) could improve the error message. After all, it's still experimental.

**Reply by BertramYe** [reply]
sure, as when I use the hooks（including the zustand） in the React 18 without starts with the use, no error occured or warning alert, who knows ┑(￣Д ￣)┍.......

**Reply by RusithTharindu** [reply]
1. If you want to install zustand in react 19, Just use the following command.

`npm install zustand --legacy-peer-deps`

--legacy-peer-deps is used to install dependencies even if there are dependency conflicts.

**Comment by mhaddi12** [comment]
I am creating react app using vite which goes 
```shell
npm create vite projectName
``` 
i am using zustand when make variable of zustand My UI Goes Blank why?

**Reply by BertramYe** [reply]
hello @mhaddi12 , in my opptions, you'd better show out your codes , so that we can do you a favor...
